src/metabase/pulse/render/common.clj

```diff
@@ -22,11 +26,61 @@
   Object
   (toString [_] num-str))
 
+(defn number-formatter
+  "Return a function that will take a number and format it according to its column viz settings. Useful to compute the
+  format string once and then apply it over many values."
+  ([column col-viz]
+   (number-formatter column
+                     col-viz
+                     (mb.viz/db->norm-column-settings-entries
+                      (:type/Number (public-settings/custom-formatting)))))
+  ([{:keys [effective_type base_type] :as _column} col-viz site-settings]
+   (let [{::mb.viz/keys [number-separators decimals scale number-style
+                         prefix suffix currency-style currency]} (merge site-settings col-viz)
+
+         currency?          (boolean (or (= number-style "currency")
+                                         (and (nil? number-style)
+                                              (or
+                                               currency-style
+                                               currency))))
+         [decimal grouping] (or number-separators ".,")
+         symbols            (doto (DecimalFormatSymbols.)
+                              (cond-> decimal (.setDecimalSeparator decimal))
+                              (cond-> grouping (.setGroupingSeparator grouping)))
+         base               (if (= number-style "scientific") "0" "#,###")
+         fmt-str            (cond-> (cond
+                                      decimals  (apply str base "." (repeat decimals "0"))
+
+                                      (isa? (or effective_type base_type) :type/Integer)
+                                      base
+
+                                      :else     (str base ".00"))
+                              (= number-style "scientific") (str "E0")
+                              (= number-style "percent")    (str "%"))
+         fmtr               (DecimalFormat. fmt-str symbols)]
+     (fn [value]
+       (if (number? value)
+         (NumericWrapper.
+          (str (when prefix prefix)
+               (when (and currency? (or (nil? currency-style)
+                                        (= currency-style "symbol")))
+                 (get-in currency/currency [(keyword (or currency "USD")) :symbol]))
+               (when (and currency? (= currency-style "code"))
+                 (str (get-in currency/currency [(keyword (or currency "USD")) :code]) \space))
+               (.format fmtr (* value (or scale 1)))
+               (when (and currency? (= currency-style "name"))
+                 (str \space (get-in currency/currency [(keyword (or currency "USD")) :name_plural])))
```

Comment:
On the frontend, the key `mv.viz/currency-in-header` key determines whether a currency value in a table should be formatted -- it should only be explicitly formatted when it is `false`, not when it's `true` or omitted.

Unfortunately it seems that in chart axes, currency values are always formatted regardless of this key. So I'm not sure if there's a way to get this right unless this code has context about whether a value will be rendered in a table or a chart. 😕